health.go

```diff
@@ -242,25 +286,57 @@ func (h *Health) startRunner(cfg *Config, ticker *time.Ticker, stop <-chan struc
 
 		h.Logger.WithFields(log.Fields{"name": cfg.Name}).Debug("Checker exiting")
 	}()
-
-	return nil
 }
 
+// resets the states in a concurrency-safe manner
 func (h *Health) safeResetStates() {
 	h.statesLock.Lock()
 	defer h.statesLock.Unlock()
 	h.states = make(map[string]State, 0)
 }
 
+// updates the check state in a concurrency-safe manner
 func (h *Health) safeUpdateState(stateEntry *State) {
 	// update states here
 	h.statesLock.Lock()
 	defer h.statesLock.Unlock()
+
+	// dispatch any status listeners
+	h.handleStatusListener(stateEntry)
+
 	h.states[stateEntry.Name] = *stateEntry
 }
 
+// get all states in a concurrency-safe manner
 func (h *Health) safeGetStates() map[string]State {
 	h.statesLock.Lock()
 	defer h.statesLock.Unlock()
 	return h.states
 }
+
+// if a status listener is attached
+func (h *Health) handleStatusListener(stateEntry *State) {
+	if h.StatusListener == nil {
+		return
+	}
+
+	// get the previous state
+	prevState := h.states[stateEntry.Name]
+
+	// state is failure
+	if stateEntry.isFailure() {
+		if !prevState.isFailure() {
+			// new failure: previous state was ok
+			go h.StatusListener.HealthCheckFailed(stateEntry)
+			stateEntry.timeOfFirstFailure = time.Now()
+		} else {
+			// carry the time of first failure from the previous state
+			stateEntry.timeOfFirstFailure = prevState.timeOfFirstFailure
+		}
+		stateEntry.contiguousFailures = prevState.contiguousFailures + 1
+	} else if prevState.isFailure() {
+		// recovery, previous state was failure
+		failureSeconds := time.Now().Sub(prevState.timeOfFirstFailure).Seconds()
+		go h.StatusListener.HealthCheckRecovered(stateEntry, prevState.contiguousFailures, failureSeconds)
+	}
```

Comment:
Since this PR has been sitting in review for this long, I think we should change the `handleStatusListener` order (as suggested by you) and get this merged. If we decide we want to expose the metadata, it can be done in a separate PR (without any backwards compat breakage).